src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -352,64 +396,506 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+  last_part = true;
+  return handle_data(bl_rem, 0, bl_rem.length());
+}
+
+int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+{
+  auto rgw_get_obj_max_req_size = g_conf()->rgw_get_obj_max_req_size;
+
+  if (!last_part && bl.length() <= rgw_get_obj_max_req_size) {
+    auto r = client_cb->handle_data(bl, bl_ofs, bl_len);
+
+    if (r < 0) {
+      return r;
+    }
+  }
+
+  int ret = 0;
+  //Accumulating data from backend store into rgw_get_obj_max_req_size sized chunks and then writing to cache
+  if (write_to_cache) {
+    const std::lock_guard l(d3n_get_data.d3n_lock);
+    rgw::d4n::CacheBlock block;
+    rgw::d4n::BlockDirectory* blockDir = source->driver->get_block_dir();
+    if (bl.length() > 0 && last_part) { // if bl = bl_rem has data and this is the last part, write it to cache
+      std::string oid = this->oid + "_" + std::to_string(ofs) + "_" + std::to_string(bl_len);
+      ret = filter->get_cache_driver()->put_async(save_dpp, oid, bl, bl.length(), source->get_attrs());
+      if (ret == 0) {
+        block.size = bl.length();
+        block.blockId = oid;
+        block.hostsList.push_back(blockDir->get_addr().host + ":" + std::to_string(blockDir->get_addr().port)); //check if this is the correct way to get the local addr
+        block.size = source->get_obj_size();
+        block.cacheObj.bucketName = source->get_bucket()->get_name();
+        block.cacheObj.objName = source->get_key().get_oid();
+        if (filter->get_policy_driver()->cachePolicy->get_block(save_dpp, &block, filter->get_cache_driver()) == 0) {
+          filter->get_policy_driver()->cachePolicy->insert(save_dpp, oid, ofs, bl.length(), filter->get_cache_driver());
+        }
+        /* Store block in directory */
+        if (!blockDir->exist_key(oid)) {
+          int ret = blockDir->set_value(&block);
+          if (ret < 0) {
+            ldpp_dout(save_dpp, 0) << "D4N Filter: Block directory set operation failed." << dendl;
+            return ret;
+          } else {
+            ldpp_dout(save_dpp, 20) << "D4N Filter: Block directory set operation succeeded." << dendl;
+          }
+	      }
+      }
+    } else if (bl.length() == rgw_get_obj_max_req_size && bl_rem.length() == 0) { // if bl is the same size as rgw_get_obj_max_req_size, write it to cache
+      std::string oid = this->oid + "_" + std::to_string(ofs) + "_" + std::to_string(bl_len);
+      ofs += bl_len;
+      block.blockId = oid;
+      block.hostsList.push_back(blockDir->get_addr().host + ":" + std::to_string(blockDir->get_addr().port)); //check if this is the correct way to get the local addr
+      block.size = source->get_obj_size();
+      block.cacheObj.bucketName = source->get_bucket()->get_name();
+      block.cacheObj.objName = source->get_key().get_oid();
+      ret = filter->get_cache_driver()->put_async(save_dpp, oid, bl, bl.length(), source->get_attrs());
+      if (ret == 0) {
+        block.size = bl.length();
```

Comment:
@samarahu : actually the size was set here correctly, the previous statement needs to be just removed.